### PR TITLE
Re-add initial Flash Attention support on ROCM

### LINKED ISSUE #112997: Add support for Flash Attention for AMD/ROCm
### 🚀 The feature, motivation and pitch

Enable support for Flash Attention Memory Efficient and SDPA kernels for AMD GPUs. 

At present using these gives below warning with latest nightlies (torch==2.2.0.dev20231105+rocm5.6, pytorch-triton-rocm==2.1.0+34f8189eae): 

> model.py:187: UserWarning: 1Torch was not compiled with flash attention. (Triggered internally at ../aten/src/ATen/native/transformers/hip/sdp_utils.cpp:253.)
  y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0)
model.py:187: UserWarning: 1Torch was not compiled with memory efficient attention. (Triggered internally at ../aten/src/ATen/native/transformers/hip/sdp_utils.cpp:291.)
  y = F.scaled_dot_product_attention(q, k, v, attn_mask=mask, dropout_p=0.0)
  
ROCm already has an implementation of Tri's FA here: https://github.com/ROCmSoftwarePlatform/flash-attention/tree/flash_attention_for_rocm2#amd-gpurocm-support

### Alternatives

User's have to manually install the ROCm version of FA and use that in their code, vs using the native PyTorch APIs. 

### Additional context

The ROCM build currently has the FA related flags turned off by default: https://github.com/pytorch/pytorch/blob/main/CMakeLists.txt#L741-L750

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### PR DESCRIPTION
**Pull request by xinyazhang**
Note about the Updates:

This PR:
1. skips more flash attention related UTs on MI200
2. Fix additional ATen compiling errors after hipification
3. Fix the author "root" of a specific commit
4. Includes the patch from Nikita in favor of block level static initialization.

CAVEAT: This revised PR has a commit that modifies the CI to force its running on MI200 nodes. That specific commit must be reverted before merge.

Original PR (https://github.com/pytorch/pytorch/pull/114309) Note:

This pull requests add initial Flash Attention support for AMD/ROCM platform. It added a specialized Triton repository/branch as a compile-time dependency for Flash Attention math library on AMD/ROCM. This triton submodule is not used at runtime and will not be shipped to the final pytorch package. We have the plan to release this specialized Triton as a separate project.

Know limitations:

- Only supports MI200 series GPU (i.e., `gcnArchName == gfx90a:sramecc+:xnack-`.
- Only supports power of two sequence lengths.
- No support for varlen APIs.
- Only support head dimension 16,32,64,128.
- Performance is still being optimized.

Fixes #112997

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

### CODE REVIEW

**Comment by xinyazhang** [comment]
Note to maintainer.

This PR has a commit that modifies the CI to force it running on MI200 nodes. This commit must be removed before merge.

**Comment by xinyazhang** [comment]
Again: DO NOT merge this without dropping commit https://github.com/pytorch/pytorch/pull/115981/commits/d028c5b7b8136cd38f478cf1a458a28d2d5c961c

File: caffe2/CMakeLists.txt
Comment (malfet) [inline caffe2/CMakeLists.txt:958]: Shouldn't this be guarded by `if(USE_FLASH_ATTENTION)`?

File: cmake/Dependencies.cmake
Comment (malfet) [inline cmake/Dependencies.cmake:1298]: Should this be guarded by `if(USE_FLASH_ATTENTION)`

File: .github/workflows/rocm.yml
Comment (malfet) [inline .github/workflows/rocm.yml:33]: From the PR description, it looks like this change should be reverted, shouldn't it?

Comment (jeffdaily) [reply .github/workflows/rocm.yml:33]: It will be, yes.  IIUC we've confirmed the MI210s have given a green signal now, so we will revert this change.

**Review by malfet — APPROVED** [review]

**Comment by xinyazhang** [comment]
@malfet Wait a sec I haven't reverted the MI210 CI change.

File: caffe2/CMakeLists.txt
Comment (xinyazhang) [reply caffe2/CMakeLists.txt]: @malfet After trying for a few more cycles, no, we can't, unfortunately.

`USE_FLASH_ATTENTION` and `USE_MEM_EFF_ATTENTION` are used in `cmake/Dependencies.cmake` to enable FA/MEFF, but these two flags depends on `USE_ROCM`, which defaults to True and will be set correctly inside `cmake/Dependencies.cmake`.

**Comment by jithunnair-amd** [comment]
All unit tests passed on MI210 in `rocm` workflow:
https://hud.pytorch.org/pytorch/pytorch/pull/115981?sha=1a76dbeb5549f809d8d72355573a17e65f44f342

**Pull request closed without merging**
